### PR TITLE
feat: support import attributes with `ImportAttribute`

### PR DESCRIPTION
**Pull request by aralroca**
Fixes https://github.com/meriyah/meriyah/issues/253

This PR implements support for Import Attributes, aligning with the [Import Attributes proposal](https://github.com/tc39/proposal-import-attributes) and [Estree import attributes](https://github.com/estree/estree/blob/master/stage3/import-attributes.md).

### CODE REVIEW

**Comment by 3cp** [comment]
Great work!

From your implementation I noticed following syntax will work:
```js
// trailing comma
import foo from "foo" with { type: "json", };
import("foo", {type: "module", });

// and some expression features like following.
import("foo", {[type = "a"]: "module"});
import("foo", {type: () => {}});
// some expression features can be rejected by playing Context and canAssign and inGroup.
```

I am not clear whether the proposal allows them, the proposal didn't have a detailed parsing algorithm like normal ecmascript spec provided. But we can probably leave with them for now. Suggest to add few more test lines to the pass/fail list.

Another thing is you need to remove `import-assertions` and `import-attributes` from `test/test262-parser-tests/unsupported-features.txt`. That will turn some more test262 test cases, sure they will give you some more work to do :-) For some you cannot figure out solution, the last thing we can do is to add some cases to `test/test262-parser-tests/whitelist.txt`. The whitelist are the cases meriyah has issue with, we do have a huge list of problems.

Happy to merge your work after passing (most :-)) test262 cases.

**Comment by 3cp** [comment]
Ignore the nodejs v22 failure. v22.5.0 has some issue right now.

**Comment by 3cp** [comment]
Many are in the whitelist right now, you need to remove them too.

**Comment by 3cp** [comment]
@fisker please help.
For this branch, I removed `import-assertions` and `import-attributes` from `unsupported-features.txt`, and various entries in `whitelist.txt`.

But the test262 summary didn't change anything before and after my edit?? Any other thing is controlling test262 cases?

```
huocp@Chunpengs-MacBook-Pro ~/p/meriyah (master)> bun run test
$ mocha --"test/**/*.ts"

  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬]Testing complete.
Summary:
 ✔ 81974 valid programs parsed without error
 ✔ 7479 invalid programs produced a parsing error
 ✔ 454 invalid programs did not produce a parsing error (and allowed by the whitelist file)
 ✔ 1843 valid programs produced a parsing error (and allowed by the whitelist file)
 ✔ 1674 programs were skipped
  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬]

  93947 passing (12s)

huocp@Chunpengs-MacBook-Pro ~/p/meriyah (master)> bun run test
$ mocha --"test/**/*.ts"

  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬]Testing complete.
Summary:
 ✔ 81974 valid programs parsed without error
 ✔ 7479 invalid programs produced a parsing error
 ✔ 454 invalid programs did not produce a parsing error (and allowed by the whitelist file)
 ✔ 1843 valid programs produced a parsing error (and allowed by the whitelist file)
 ✔ 1674 programs were skipped
  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬]

  93947 passing (12s)
```

**Comment by 3cp** [comment]
@fisker ignore my deleted comment. I was editing wrong directory.

**Comment by 3cp** [comment]
There are some valid reason we need some cases in whitelist.

Like `language/import/import-attributes/json-invalid.js`, which asks the parser to read and check the JSON file. Meriyah doesn't do that and would not do that.

```js
import value from './json-invalid_FIXTURE.json' with { type: 'json' };
```

**Comment by 3cp** [comment]
Correction: import-assertions is the outdated proposal that import-attributes replaced. Please keep import-assertions in unsupported list.

**Comment by aralroca** [comment]
> There are some valid reasons we need some cases in whitelist.
> 
> Like `language/import/import-attributes/json-invalid.js`, which asks the parser to read and check the JSON file. Meriyah doesn't do that and would not do that.
> 
> ```js
> import value from './json-invalid_FIXTURE.json' with { type: 'json' };
> ```

This is the only case I added to the whitelist, the other ones are fixed! Feel free to re-review @3cp ! Thanks a lot!

File: src/parser.ts
Comment (aralroca) [inline src/parser.ts:4475]: I'm not sure about this, they pass the tests, but I was really just looking for “no crash”, but I think if I add tests of the expected AST to test262/test/language/expressions/dynamic-import/import-attributes/2nd-param-in.js it will crash. Tomorrow I will continue and add more tests from test262 with the AST after parsing.

File: src/parser.ts
Comment (3cp) [inline src/parser.ts:4483]: consumeOpt is for this kind of optional.

File: src/parser.ts
Comment (3cp) [reply src/parser.ts:4475]: Yes, this is not right.
"in" is handled by parseBinaryExpression. The reason that parseExpression didn't cover it, is probably Context.DisallowIn is turned on. Probably need just `(context | Context.DisallowIn) ^ Context.DisallowIn` to turn it off.

Comment (aralroca) [reply src/parser.ts]: Okay perfect, it looks better now. I've added the failing test with the AST and corrected it with what you told me.

File: src/parser.ts
Comment (fisker) [inline src/parser.ts:4519]: Why is this needed? The value of type shouldn't matters at all.

Comment (fisker) [reply src/parser.ts:4519]: Ah, because we support `json-modules` feature? https://github.com/tc39/proposal-json-modules#why-dont-json-modules-support-named-exports

Comment (aralroca) [reply src/parser.ts:4519]: @fisker this is to fix this test: `test262/test/language/import/import-attributes/json-named-bindings.js`. As it says:

```
In the early design of JSON modules, contributors considered allowing the
  properties of object values in JSON modules to be imported directly by name.
  This was ultimately rejected, so attempting to import in this way should
  produce a SyntaxError.
```

However, this is specific only to JSON. For other import attributes, such as Bun macros, it has to be done for named bindings: https://bun.sh/docs/bundler/macros

Comment (fisker) [reply src/parser.ts:4519]: I don't think any parser rejects this now, maybe we can put `json-modules` into [unsupported features](https://github.com/meriyah/meriyah/blob/master/test/test262-parser-tests/unsupported-features.txt). But lets hear what @3cp thinks first.

Comment (aralroca) [reply src/parser.ts:4519]: No problem, or I imagine it will be putting this `test262/test/language/import/import-attributes/json-named-bindings.js` to the whitelist. Well, I'll wait for @3cp's answer and see what he thinks. Thanks!

File: src/errors.ts
Comment (3cp) [inline src/errors.ts:142]: Just noticed you forgot to add errorMessages for these two.

Comment (3cp) [reply src/errors.ts:142]: Which causes:
```
> parseModule('import { foo } from "./foo.json" with { type: "json" };', { next: true })
/Users/huocp/playground/meriyah/src/errors.ts:176
        const message = '[' + line + ':' + column + ']: ' + exports.errorMessages[type].replace(/%(\d+)/g, (_, i) => params[i]);
                                                                                        ^

Uncaught TypeError: Cannot read properties of undefined (reading 'replace')
```

File: src/parser.ts
Comment (3cp) [inline src/parser.ts:4505]: This enables
```js
import { default as a, foo } from "./foo.json" with { type: "json" };
// and
import a, { foo } from "./foo.json" with { type: "json" };
```

It probably should be `specifiers.length === 1`.

Comment (3cp) [reply src/parser.ts:4505]: I think test262 missed some test cases for this.

File: src/parser.ts
Comment (3cp) [reply src/parser.ts:4519]: I wish tc39 pushed this as runtime error, not syntax error :-)

File: src/parser.ts
Comment (3cp) [inline src/parser.ts:4522]: The validJSONImportAttributeBindings logic can probably be delayed here. So we don't have the unnecessary cost for non json import.

File: src/parser.ts
Comment (3cp) [inline src/parser.ts:4547]: Suggest to rename it to parseIdentifierOrStringLiteral.
Do we have any test cases to reject?
```
with { 1: "foo" }
with { type: 1 }
with { type: [1] }
with { type: null }
with { type: undefined }
with { type: 'json', foo: {} }
...
```

**Review by 3cp — CHANGES_REQUESTED** [review]

File: src/errors.ts
Comment (3cp) [reply src/errors.ts:142]: Should it be named InvalidJSONImportBinding?

File: src/parser.ts
Comment (3cp) [reply src/parser.ts:4547]: Looks like your implementation didn't reject the above cases.

From the proposal: 
```
We could permit import attributes to have more complex values than simply strings, for example:

import value from "module" with { attr: { key1: "value1", key2: [1, 2, 3] } };
This would allow import attributes to scale to support a larger variety of metadata.

We propose to omit this generalization in the initial proposal, as a key/value list of strings already affords significant flexibility to start, but we're open to a follow-on proposal providing this kind of generalization.
```

Comment (3cp) [reply src/parser.ts:4547]: test262 cases are way inadequate :-)

File: src/errors.ts
Comment (aralroca) [reply src/errors.ts]: I have made the changes that you have proposed, I have also deleted the InvalidImportExpression, since in the end it is not used after the changes to adapt to test252.

**Comment by aralroca** [comment]
@3cp @fisker Thank you for all your suggestions for changes. Feel free to check the PR again, anything else feel free to say. Looking forward to have the import attributes supported 😊

**Review by 3cp — APPROVED** [review]

**Comment by 3cp** [comment]
LGTM.
Thx!
Let @fisker have another look.

**Review by fisker — APPROVED** [review]
LGTM, thank you  @aralroca

File: src/parser.ts
Comment (Tcjmd) [inline src/parser.ts]: In this case can it become a binding? Anyway the expr parser should detect and throw if missing RightBrace. Meaning this check is both expensive due to perf and not needed

File: src/parser.ts
Comment (Tcjmd) [inline src/parser.ts:4533]: This is handcoded and no AST node. No ESTree node for this?

Comment (3cp) [reply src/parser.ts:4533]: Most finishNode usage in meriyah looks like this, with plain string type.

File: src/parser.ts
Comment (3cp) [reply src/parser.ts]: That's not latest code.

Comment (Tcjmd) [reply src/parser.ts]: My mistake. Can you link to latest. On cp

Comment (3cp) [reply src/parser.ts]: Just visit the changed code in this PR, you will see latest code.

**Pull request merged**
